CPnest merge into master

nielsenmb opened the pull request:
This PR merges the CPnest bits of the branch #232  into master (there are reggae bits that are not merged). 

Optional arguments are put into session, star and asy_peakbag to pick the sampling method. 

CPnest apparently dumps a lot of stuff in cwd. This is now sent to a cpnest subdirectory in the star directory instead. 

Questions for @grd349 
- self.acceptance= np.inf in the nested class is TODO. What is this?
- In asy_peakbag, min/max of the samples from KDE are used to set the bounds of CPnest. These samples are drawn from the kde,, but in using the min/max of the sample, are we not effectively truncating the KDE prior?

grd349 commented:
'self.acceptance= np.inf in the nested class is TODO. What is this?'

I think I saw that star or session used self.acceptance so to maintain compatibility I added this.  If not needed than just delete.

'In asy_peakbag, min/max of the samples from KDE are used to set the bounds of CPnest.' 

They need not be set like this.  This was probably just easier for me to set the bounds.  We could probably just set the bounds to something universal for all stars - might need some thought as to what that is.  If the bounds are too large that won't make a big difference as long as we sample from the prior in CPnest.

grd349 commented on pbjam/mcmc.py:
```diff
@@ -235,3 +238,80 @@ def fold(self, pos, accept_lim = 0.2, spread=0.1):
+        Parameters
+        ----------
+        nlive : int
+            ???
```
Can probably just pull in from the CPnest docs here

grd349 commented on pbjam/asy_peakbag.py:
```diff
@@ -377,12 +375,23 @@ def __call__(self, developer_mode):
+            self.fit(start_samples=self.start_samples)
+
+        elif method == 'nested':
+            bounds = [[self.start_samples[:, n].min(), 
```
I think for now these bounds are necessary as the prior is not universal.  When we switch to a universal prior this will be better (I think)

nielsenmb replied:
Probably fine for now.

grd349 commented:
I was wrong about the universal bounds!  The prior is not universal so we need to limit the bounds to the extremes of the prior.

When we make the prior universal then we can change this!

nielsenmb commented:
Was just about to fix those docs strings.

On Fri, 12 Jun 2020 at 13:19, Guy Davies <notifications@github.com> wrote:

> *@grd349* approved this pull request.
>
> Looks good to me - jiust the ???? to sort.
> ------------------------------
>
> In pbjam/mcmc.py
> <https://github.com/grd349/PBjam/pull/242#discussion_r439383072>:
>
> > +        """ Wrapper for log likelihood """
> +        return self.likelihood(param.values)
> +
> +    def log_prior(self,p):
> +        """ Wrapper for log prior """
> +        if not self.in_bounds(p): return -np.inf
> +        return self.prior(p.values)
> +
> +    def __call__(self, nlive=100, nthreads=1, maxmcmc=100, poolsize=100):
> +        """
> +        Runs the nested sampling
> +
> +        Parameters
> +        ----------
> +        nlive : int
> +            ???
>
> Can probably just pull in from the CPnest docs here
> ------------------------------
>
> In pbjam/asy_peakbag.py
> <https://github.com/grd349/PBjam/pull/242#discussion_r439384281>:
>
> > @@ -377,12 +375,23 @@ def __call__(self, developer_mode):
>          """
>          self.developer_mode = developer_mode
>
> -        self.fit = pb.mcmc(np.median(self.start_samples, axis=0),
> -                           self.likelihood, self.prior)
> -        self.fit(start_samples=self.start_samples)
> +        if method not in ['mcmc', 'nested']:
> +            warnings.warn(f'Method {method} not found: Using method mcmc')
> +            method = 'mcmc'
> +
> +        if method == 'mcmc':
> +            self.fit = pb.mcmc(np.median(self.start_samples, axis=0),
> +                               self.likelihood, self.prior)
> +            self.fit(start_samples=self.start_samples)
> +
> +        elif method == 'nested':
> +            bounds = [[self.start_samples[:, n].min(),
>
> I think for now these bounds are necessary as the prior is not universal.
> When we switch to a universal prior this will be better (I think)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grd349/PBjam/pull/242#pullrequestreview-429692131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJWO36BP7OMTK6Y5ASY3VDRWIMNVANCNFSM4N4G47CQ>
> .
>

nielsenmb commented:
Fixed up the docstrings.

For the self.acceptance, it's only stored in the emcee case so the user has that info available later. Changed it to None for the nested case to avoid confusion.

nielsenmb commented:
You mean the extremes of the stars in the prior sample that are selected to
create the kde right? (numax \pm sigma etc.) I was about to ask why we
weren't doing that.

That should be considerably wider than the pdf from the kde, but probably
the situation we want to be in. Not the other way around.

On Fri, 12 Jun 2020 at 13:21, Guy Davies <notifications@github.com> wrote:

> I was wrong about the universal bounds! The prior is not universal so we
> need to limit the bounds to the extremes of the prior.
>
> When we make the prior universal then we can change this!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grd349/PBjam/pull/242#issuecomment-643241747>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJWO3ZGJGJJOL5BMARVFFDRWIMUXANCNFSM4N4G47CQ>
> .
>

nielsenmb commented:
Set the the bounds to the min/max of the data subset selected to generate the kde prior.

For Boeing the result are identical to using emcee.

grd349 commented:
Great!  Yes - let's merge.  Does that mean you should update the paper with the second method?

The CPNest folks have just merged my PR into master for the sample_prior keyword.  If you download the repo from github you could see if CPnest gets any faster with the keyword set to True.

nielsenmb commented:
I will just add a bit to say that the choice of sampler is optional and
that we currently have emcee and CPnest implemented.

On Fri, 12 Jun 2020 at 14:41, Guy Davies <notifications@github.com> wrote:

> Great! Yes - let's merge. Does that mean you should update the paper with
> the second method?
>
> The CPNest folks have just merged my PR into master for the sample_prior
> keyword. If you download the repo from github you could see if CPnest gets
> any faster with the keyword set to True.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grd349/PBjam/pull/242#issuecomment-643276610>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEJWO35LYEPDRNZFEEXART3RWIWALANCNFSM4N4G47CQ>
> .
>

grd349 commented:
perfect.  

Do you need me to merge?  (why aren't the tests working!)

G

nielsenmb commented:
My bad, I didn't update the tests properly. 

Pytest passes completely locally for me now.

nielsenmb commented:
@grd349 can you merge this now?
